tests/test_api.py

```diff
@@ -786,3 +786,163 @@ def test_product_components(client, api_path):
+        defaults={"obj": root_comp},
+    )
+
+    # create upstream child node
+    upstream_comp = ComponentFactory(
+        name="upstream_comp",
+        type=Component.Type.RPM,
+        namespace=Component.Namespace.UPSTREAM,
+        related_url="https://example.org/related",
+    )
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.SOURCE,
+        parent=root_node,
+        purl=upstream_comp.purl,
+        defaults={"obj": upstream_comp},
+    )
+
+    # create dep component
+    dep_comp = ComponentFactory(
+        name="cool_dep_component",
+        arch="aarch64",
+        type=Component.Type.RPM,
+        namespace=Component.Namespace.REDHAT,
+    )
+    # make it a child of root component
+    dep_provide_node, _= ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.PROVIDES,
+        parent=root_node,
+        purl=dep_comp.purl,
+        defaults={"obj": dep_comp},
+    )
+    # create 2nd level dep
+    dep2_comp = ComponentFactory(
+        name="cool_dep2_component", type=Component.Type.PYPI, namespace=Component.Namespace.REDHAT
+    )
+    # making it a child node of dep1_component
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.PROVIDES,
+        parent=dep_provide_node,
+        purl=dep2_comp.purl,
+        defaults={"obj": dep2_comp},
+    )
+    ComponentNode.objects.get_or_create(
+        type=ComponentNode.ComponentNodeType.PROVIDES,
+        parent=dep_provide_node,
+        purl=dep2_comp.purl,
+        defaults={"obj": dep2_comp},
+    )
+
+    dep2_comp.save_component_taxonomy()
+    dep_comp.save_component_taxonomy()
+    root_comp.save_component_taxonomy()
+    root_comp.save()
+
+    assert dep_comp.purl in root_comp.provides.values_list("purl", flat=True)
+    assert dep2_comp.purl in root_comp.provides.values_list("purl", flat=True)
+
+    response = client.get(f"{api_path}/components?namespace=REDHAT")
+    assert response.status_code == 200
+    assert response.json()["count"] == 3
+
+    response = client.get(f"{api_path}/components/{root_comp.uuid}")
+    assert response.status_code == 200
+    assert response.json()["name"] == root_comp.name
+    assert response.json()["related_url"] == root_comp.related_url
+    assert len(response.json()["sources"]) == 0
+    assert len(response.json()["provides"]) == 2
+    assert len(response.json()["upstreams"]) == 1
+
+    response = client.get(f"{api_path}/components/{dep_comp.uuid}")
+    assert response.status_code == 200
+    assert len(response.json()["sources"]) == 1
+    assert len(response.json()["provides"]) == 1
+    assert len(response.json()["upstreams"]) == 1
+
+    response = client.get(f"{api_path}/components/{dep2_comp.uuid}")
+    assert response.status_code == 200
+    assert len(response.json()["sources"]) == 2
+    assert len(response.json()["provides"]) == 0
+    assert len(response.json()["upstreams"]) == 1
```

Comment:
I think this was correct in the first version I reviewed, unless I've misunderstood. Today, only the root component reports an upstream component.
```suggestion
    assert len(response.json()["upstreams"]) == 0
```